core/src/main/java/org/elasticsearch/action/fieldstats/TransportFieldStatsTransportAction.java

```diff
@@ -208,13 +212,19 @@ protected FieldStatsShardResponse shardOperation(FieldStatsShardRequest request)
                         // ignore.
                         continue;
                     }
-                    FieldStats<?> stats = fieldType.stats(searcher.reader());
-                    if (stats != null) {
-                        fieldStats.put(matchField, stats);
+                    BytesReference cacheKey = new BytesArray("fieldstats:" + matchField);
+                    BytesReference statsRef = indicesService.cacheShardLevelResult(shard, searcher.getDirectoryReader(), cacheKey, out -> {
+                        out.writeOptionalWriteable(fieldType.stats(searcher.reader()));
+                    });
+                    try (StreamInput in = StreamInput.wrap(statsRef)) {
+                        FieldStats<?> stats = in.readOptionalWriteable(FieldStats::readFrom);
```

Comment:
should we fix `FieldStatsShardResponse` to take the opaque representation per field? I think it would make it less complex and would safe so reading / writing?